### PR TITLE
Adding MS-MPI

### PR DESCRIPTION
**Pull request by RyanMcCarthy-NOAA**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by isuruf** [comment]
Try removing the line https://github.com/microsoft/Microsoft-MPI/blob/master/Directory.Build.props#L117 or changing the value to `14.16.27023`

**Comment by RyanMcCarthy-NOAA** [comment]
@isuruf that might work locally, but I'll have to ask the msmpi devs to make that change.

**Comment by isuruf** [comment]
You can add it as a patch. See https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#patches

**Comment by RyanMcCarthy-NOAA** [comment]
@isuruf thanks for the tip, that helped a lot! Looks like I also need to set the gfortran path in that file as well. I tried `$(LIBRARY_BIN)`, which ends up as `%PREFIX%\Library\bin\gfortran.exe` for gfortran, but that fails. Do you what the right env-var/path to use for this is?

**Comment by leofang** [comment]
Thanks for working on this, @RyanMcCarthy-NOAA!

cc: @dalcinl

File: recipes/msmpi/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  number: 0
+  skip: true  # [not win]
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ compiler('c') }}
+    - {{ compiler('m2w64_fortran') }}
+  host:
+  run:
```
Comment (scopatz) [inline recipes/msmpi/meta.yaml:24]: please remove empty sections

Comment (RyanMcCarthy-NOAA) [reply recipes/msmpi/meta.yaml]: Done!

**Comment by RyanMcCarthy-NOAA** [comment]
Thanks to @isuruf for the assistance, but I think the full path to gfortran is still not correct. @scopatz , any advice getting that defined? Currently using:
```
$(LIBRARY_PREFIX)\mingw-w64\bin
```
but that is returning `9009` errors, which I think is a file not found error?

**Comment by scopatz** [comment]
$FC should be the path to the fortran compiler

**Comment by isuruf** [comment]
> $FC should be the path to the fortran compiler

No, it's not. Please see https://github.com/conda-forge/staged-recipes/pull/12506#discussion_r479679185

**Comment by scopatz** [comment]
FC should probably be set by the activation scripts then

**Comment by RyanMcCarthy-NOAA** [comment]
@isuruf , looks like you got the build working (awesome work!) but I left a test in there that is failing. Again, I'm unfamiliar with the msbuild process, so wasn't sure what/where to check for build outputs. Do you know if there is an "install" target with this project? I could see builds going to `%SRC_DIR%\out\Release-x64` and `%SRC_DIR%\out\Release-x64\bin` but I'm not sure if the correct thing to do is directly copy from there to `LIBRARY_PREFIX` or something else.

**Comment by isuruf** [comment]
I don't think there is an install code and you are right that we have to copy from `%SRC_DIR%\out\Release-x64\` to `%LIBRARY_PREFIX%`

File: recipes/taudem/build.sh [hunk 2]
```diff
@@ -1,8 +0,0 @@
-#!/bin/bash
```
Comment (scopatz) [inline recipes/taudem/build.sh:1]: Pelase merge from master

**Comment by isuruf** [comment]
Some files are not in the correct location.
1. All `.exe`s and `.dll`s need to be in `%LIBRARY_PREFIX%\bin`
2. All `.h`s and `.f`s need to be in `%LIBRARY_PREFIX%\include`
3. All `.lib`s need to be in `%LIBRARY_PREFIX%\include`

File: recipes/msmpi/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ compiler('c') }}
+    - {{ compiler('m2w64_fortran') }}
+
+test:
+  commands:
+    - if not exist %LIBRARY_PREFIX%\\bin\\msmpi.dll exit 1  # [win]
```
Comment (isuruf) [inline recipes/msmpi/meta.yaml:26]: Can you add a test for one file of each of `.exe, .h, .f, .lib` files?

Comment (RyanMcCarthy-NOAA) [reply recipes/msmpi/meta.yaml:26]: that's a good idea. I'll add that.

File: recipes/msmpi/bld.bat [hunk 4]
```diff
@@ -9,15 +9,18 @@ msbuild.exe /p:Platform=%PLATFORM% /p:Configuration=Release
...
-robocopy %SRC_DIR%\out\Release-x64\bin\sdk\inc\pmidbg.h %LIBRARY_INC%\msmpi\pmidbg.h /s /e
-robocopy %SRC_DIR%\out\Release-x64\bin\sdk\inc\x64\mpifptr.h %LIBRARY_INC%\msmpi\x64\mpifptr.h /s /e
-robocopy %SRC_DIR%\out\Release-x64\bin\sdk\inc\x86\mpifptr.h %LIBRARY_INC%\msmpi\x86\mpifptr.h /s /e
+mkdir %LIBRARY_INC%\msmpi
+mkdir %LIBRARY_INC%\msmpi\x64
+
+copy %SRC_DIR%\out\Release-x64\bin\sdk\inc\mpi.f90 %LIBRARY_INC%\msmpi\mpi.f90
+copy %SRC_DIR%\out\Release-x64\bin\sdk\inc\mpi.h %LIBRARY_INC%\msmpi\mpi.h
+copy %SRC_DIR%\out\Release-x64\bin\sdk\inc\mpif.h %LIBRARY_INC%\msmpi\mpif.h
+copy %SRC_DIR%\out\Release-x64\bin\sdk\inc\mpio.h %LIBRARY_INC%\msmpi\mpio.h
```
Comment (isuruf) [inline recipes/msmpi/bld.bat:18]: These should go in `%LIBRARY_INC%` instead of `%LIBRARY_INC%\msmpi`

Comment (RyanMcCarthy-NOAA) [reply recipes/msmpi/bld.bat:18]: lol, I'll get it right eventually :rofl: 
Do you know if that should be the same for the contents in the `x86` subdirectory? everything going into `%LIBRARY_INC%`?

Comment (isuruf) [reply recipes/msmpi/bld.bat:18]: Yes, all `.h` files should go in `%LIBRARY_INC%`.

**Comment by isuruf** [comment]
@RyanMcCarthy-NOAA, we also need files under `%SRC_DIR%\out\Release-x64\` not just `%SRC_DIR%\out\Release-x64\bin`.
For eg: `%SRC_DIR%\out\Release-x64\mpifort\mpifort.lib`

File: recipes/msmpi/bld.bat [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    set PLATFORM=Win32
+) else (
+    set PLATFORM=x64
+)
+
+msbuild.exe /p:Platform=%PLATFORM% /p:Configuration=Release
+
+for /r %SRC_DIR%\out\Release-%PLATFORM% %%f in (*.exe) do @copy "%%f" %LIBRARY_BIN%
+for /r %SRC_DIR%\out\Release-%PLATFORM% %%f in (*.dll) do @copy "%%f" %LIBRARY_BIN%
+for /r %SRC_DIR%\out\Release-%PLATFORM% %%f in (*.lib) do @copy "%%f" %LIBRARY_LIB%
```
[the remark this answers is not in the document]
Comment (RyanMcCarthy-NOAA) [reply recipes/msmpi/bld.bat:11]: Good thinking, just added that in.

**Comment by isuruf** [comment]
Thanks

**Comment by RyanMcCarthy-NOAA** [comment]
@isuruf , thank you! I appreciate all the help getting this dialed in!

**Comment by leofang** [comment]
Awesome work, @RyanMcCarthy-NOAA @isuruf!

**Pull request merged**